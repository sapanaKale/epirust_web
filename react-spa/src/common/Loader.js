import React from 'react';

export default function Loader() {
    return (
        <div id="loader" data-testid="loader" className="multi-spinners">
            <div className="spinner-grow text-dark" />
            <div className="spinner-grow text-dark" />
            <div className="spinner-grow text-dark" />
            <div className="spinner-grow text-dark" />
            <div className="spinner-grow text-dark" />
            <div className="spinner-grow text-dark" />
        </div>
    );
}