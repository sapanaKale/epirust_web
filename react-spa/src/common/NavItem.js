/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {Link, useRouteMatch} from "react-router-dom";
import React from "react";

export function NavItem({ name, linksTo, activeOnExactMatch = false }) {
  const match = useRouteMatch({
    path: linksTo,
    exact: activeOnExactMatch
  });

  return (
    <li className={`nav-item`}>
      <Link className={`nav-link ${match ? "active" : ""}`} to={linksTo}>{name}</Link>
    </li>
  );
}