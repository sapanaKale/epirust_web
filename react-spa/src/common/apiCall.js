import config from "../config";

export const post = (url, data) => {
  return fetch(`${config.API_HOST}/api${url}`, {
    method: 'POST',
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify(data)
  })
    .then(response => handleNon200(response))
}

export const get = (url) => {
  return fetch(`${config.API_HOST}/api${url}`, {
    method: 'GET',
    headers: {'Content-Type': 'application/json'}
  })
    .then(response => handleNon200(response))
}

const handleNon200 = (response) => {
  if (!response.ok) {
    throw Error(response.statusText);
  }
  return response;
}