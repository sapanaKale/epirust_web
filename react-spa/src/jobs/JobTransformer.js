/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

const extractStatus = (simulations) => {
  if(simulations.some(s => s.status === 'failed')) return 'failed'
  if(simulations.some(s => s.status === 'running')) return 'running'
  if(simulations.every(s => s.status === 'finished')) return 'finished'
  if(simulations.some(s => s.status === 'finished')) return 'running'
  return 'in-queue'
}

export const reduceStatus = (data) => {
  return {
    ...data,
    status: extractStatus(data.simulations)
  }
}
