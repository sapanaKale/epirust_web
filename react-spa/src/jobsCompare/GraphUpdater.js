/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
import io from "socket.io-client";
import config from "../config";


export default class GraphUpdater {
  #updateBuffer
  #jobData1
  #jobData2
  #lastSent
  #stopAll

  constructor(updateBuffer, jobId1, jobId2) {
    this.#updateBuffer = updateBuffer;

    this.#jobData1 = new JobTimeSeriesData(jobId1)
    this.#jobData2 = new JobTimeSeriesData(jobId2)
    this.#lastSent = 0;
    this.#stopAll = false;
  }

  start() {
    this.#startJob(this.#jobData1);
    this.#startJob(this.#jobData2);
  }

  stop() {
    this.#jobData1.closeSocket();
    this.#jobData2.closeSocket();
  }

  #checkConsistencyAndSend = (sendLimit) => {
    const start = this.#lastSent + 1;
    const end = start + sendLimit - 1;
    const isConsistent = this.#checkConsistency(end)

    if (isConsistent) {
      const data = this.#buildData(start, end)
      if(!this.#stopAll) {
        this.#updateBuffer(data)
        this.#lastSent = end;
      }
    }
  }

  #startJob = (jobData) => {
    const socket = io(`${config.API_HOST}/counts`)
    jobData.setSocket(socket);
    socket.on('epidemicStats', (message) => {
      if ("simulation_ended" in message) {
        jobData.fetchFinished = true;
        if (this.#bothJobsFinished())
          this.#checkConsistencyAndSend(jobData.consumedCount() - this.#lastSent)
        socket.close()
      } else {
        jobData.insertData(message.hour, message)
        if (jobData.consumedCount() % config.RENDER_COUNT === 0) {
          this.#checkConsistencyAndSend(config.RENDER_COUNT)
        }
      }
    })
    socket.emit('get', {jobId: jobData.jobId});
  }

  #bothJobsFinished = () => {
    return this.#jobData1.fetchFinished && this.#jobData2.fetchFinished;
  }

  #checkConsistency = (till) => {
    const eitherJobsNotConsistent = this.#jobData1.consumedCount() < till || this.#jobData2.consumedCount() < till
    return !(!this.#bothJobsFinished() && eitherJobsNotConsistent);
  }

  #buildData = (from, till) => {
    const data = []
    for (let i = from; i <= till; i++) {
      data.push({'hour': i, 'job1': this.#jobData1.dataBuffer[i], 'job2': this.#jobData2.dataBuffer[i]})
    }
    return data;
  }
}

class JobTimeSeriesData {
  #consumedMessageCount
  #socket

  constructor(jobId) {
    this.jobId = jobId
    this.fetchFinished = false;
    this.dataBuffer = {}
    this.#consumedMessageCount = 0;
  }

  insertData = (hour, message) => {
    this.dataBuffer[hour] = message
    this.#consumedMessageCount += 1
  }

  consumedCount = () => this.#consumedMessageCount

  setSocket = (socket) => this.#socket = socket;

  closeSocket = () => this.#socket.close();
}
