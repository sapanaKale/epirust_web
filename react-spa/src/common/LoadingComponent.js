/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import {LOADING_STATES} from "./constants";
import Loader from "./Loader";
import React from "react";
import PropTypes from 'prop-types'

export const LoadingComponent = (props) => {
  switch (props.loadingState) {
    case LOADING_STATES.FINISHED:
      return props.children;

    case LOADING_STATES.FAILED:
      return <div>Failed to load</div>

    default:
      return <Loader/>
  }
}

LoadingComponent.propTypes = {
  loadingState: PropTypes.oneOf([1, 2, 3]).isRequired
}