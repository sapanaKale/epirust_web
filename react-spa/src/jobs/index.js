/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from "react";
import '../styles/components/jobs/_jobs-list.scss'
import {LOADING_STATES} from "../common/constants";
import {JobsContainer} from "./JobsContainer";
import {LoadingComponent} from "../common/LoadingComponent";
import {JobsRefresher} from "./JobsRefresher";

export const JobsView = () => {
  const [loadingState, updateLoadingState] = useState(LOADING_STATES.LOADING);
  const [jobs, updateJobs] = useState([]);

  useEffect(() => {
    const jobsRefresher = new JobsRefresher(updateJobs, updateLoadingState);
    jobsRefresher.start()
  }, []);

  return <LoadingComponent loadingState={loadingState}>
    <JobsContainer jobs={jobs}/>
  </LoadingComponent>;
};
