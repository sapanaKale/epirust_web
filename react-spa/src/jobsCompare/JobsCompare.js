/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React, {useEffect, useState} from "react";
import ComparerDropdowns from "./ComparerDropdowns";
import {get} from "../common/apiCall";
import {LoadingComponent} from "../common/LoadingComponent";
import {LOADING_STATES} from "../common/constants";
import {reduceStatus} from "../jobs/JobTransformer";
import {CompareView} from "./CompareView";

export default function JobsCompare() {
  const [jobs, updateJobs] = useState([]);
  const [loadingState, updateLoadingState] = useState(LOADING_STATES.LOADING);
  const [selectedJobs, updateSelectedJobs] = useState(null);

  useEffect(() => {
    get('/jobs')
      .then((res) => res.json())
      .then((jobsResponse) => {
        updateJobs(jobsResponse.map(reduceStatus).filter(j => j.status === 'finished'));
        updateLoadingState(LOADING_STATES.FINISHED)
      })
      .catch(() => updateLoadingState(LOADING_STATES.FAILED))
  }, []);

  const onCompare = (jobPair) => {
    updateSelectedJobs({
      job1: jobs.find(x => x._id === jobPair.job1),
      job2: jobs.find(x => x._id === jobPair.job2)
    })
  };

  return (
    <div>
      <LoadingComponent loadingState={loadingState}>
        <div className='row justify-content-md-center'>
          <div className='col-7'>
            <ComparerDropdowns jobs={jobs} onCompare={onCompare}/>
          </div>
        </div>
        {selectedJobs && <CompareView selectedJobs={selectedJobs}/>}
      </LoadingComponent>
    </div>
  )
}
