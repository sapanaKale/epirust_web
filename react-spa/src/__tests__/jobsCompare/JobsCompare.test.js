/*
 * EpiRust
 * Copyright (c) 2020  ThoughtWorks, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import React from "react";
import ShallowRenderer from 'react-test-renderer/shallow';
import {fireEvent, render} from '@testing-library/react'
import JobsCompare from "../../jobsCompare/JobsCompare";
import {get} from '../../common/apiCall'
import {flushPromises} from "../helper/promiseHelper";
import {act} from "react-dom/test-utils";
import GraphUpdater from '../../jobsCompare/GraphUpdater'
import {reduceStatus} from "../../jobs/JobTransformer";

jest.mock('../../common/apiCall')
jest.mock('../../jobsCompare/GraphUpdater')
jest.mock('../../jobs/JobTransformer')

describe('Jobs Compare', function () {
  const jobs = [
    {_id: '1', status: 'finished', simulations: [{_id: "345678iugf"}]},
    {_id: '2', status: 'finished', simulations: [{_id: "45678iugjj"}]}
  ];

  it('should render with default state', function () {
    const renderer = new ShallowRenderer();
    renderer.render(<JobsCompare/>)

    expect(renderer.getRenderOutput()).toMatchSnapshot()
  });

  it('should render with jobs fetched that are finished', async function () {
    const jobs = [
      {_id: 1, status: 'finished'},
      {_id: 2, status: 'running'},
      {_id: 3, status: 'in-queue'},
      {_id: 4, status: 'finished'},
      {_id: 5, status: 'failed'}
    ];
    get.mockResolvedValueOnce({json: jest.fn().mockResolvedValueOnce(jobs)})
    const {container} = render(<JobsCompare/>)

    await act(async () => {
      await flushPromises()
    })

    const dropdowns = container.querySelectorAll('select')
    expect(dropdowns).toMatchSnapshot()
  });

  it('should display failed message when fetching jobs has failed', async function () {
    get.mockRejectedValue()
    const {container} = render(<JobsCompare/>)

    await act(async () => {
      await flushPromises()
    })

    expect(container).toMatchSnapshot()
  });

  it('should start jobs fetcher on compare click', async () => {
    get.mockResolvedValueOnce({json: jest.fn().mockResolvedValueOnce(jobs)})
    const mockStart = jest.fn();
    GraphUpdater.mockImplementation(() => ({'start': mockStart, 'stop': jest.fn()}))
    const {container} = render(<JobsCompare/>)

    await act(async () => await flushPromises())

    const dropdowns = container.querySelectorAll('.form-control');
    fireEvent.change(dropdowns[0], {target: {value: '1'}})
    fireEvent.change(dropdowns[1], {target: {value: '2'}})
    fireEvent.click(container.querySelector('button'))

    expect(GraphUpdater).toHaveBeenCalledTimes(1)
    expect(GraphUpdater).toHaveBeenCalledWith(expect.any(Function), '1', '2')
    expect(mockStart).toHaveBeenCalledTimes(1)
  });

  it.skip('should update graph when graph updater updates value', async () => {
    get.mockResolvedValueOnce({json: jest.fn().mockResolvedValueOnce(jobs)})
    const mockStart = jest.fn();
    GraphUpdater.mockImplementation(() => ({'start': mockStart, 'stop': jest.fn()}))
    const {container} = render(<JobsCompare/>)

    await act(async () => await flushPromises())

    const dropdowns = container.querySelectorAll('.form-control');
    fireEvent.change(dropdowns[0], {target: {value: '1'}})
    fireEvent.change(dropdowns[1], {target: {value: '2'}})
    fireEvent.click(container.querySelector('button'))

    const updateFunction = GraphUpdater.mock.calls[0][0]
    act(() => {
      updateFunction([{hour: 1, job1: null, job2: null}])
    })
    expect(container).toMatchSnapshot()
  });

  beforeEach(() => {
    reduceStatus.mockImplementation(data => data)
  });

  afterEach(() => {
    jest.clearAllMocks();
  })
});
