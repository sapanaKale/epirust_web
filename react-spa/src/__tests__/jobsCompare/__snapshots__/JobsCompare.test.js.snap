// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jobs Compare should display failed message when fetching jobs has failed 1`] = `
<div>
  <div>
    <div>
      Failed to load
    </div>
  </div>
</div>
`;

exports[`Jobs Compare should render with default state 1`] = `
<div>
  <LoadingComponent
    loadingState={1}
  >
    <div
      className="row justify-content-md-center"
    >
      <div
        className="col-7"
      >
        <ComparerDropdowns
          jobs={Array []}
          onCompare={[Function]}
        />
      </div>
    </div>
  </LoadingComponent>
</div>
`;

exports[`Jobs Compare should render with jobs fetched that are finished 1`] = `
NodeList [
  <select
    class="form-control"
    id="job1"
    name="job1"
    required=""
  >
    <option
      disabled=""
      value=""
    >
       -- select a job -- 
    </option>
    <option>
      1
    </option>
    <option>
      4
    </option>
  </select>,
  <select
    class="form-control"
    id="job2"
    name="job2"
    required=""
  >
    <option
      disabled=""
      value=""
    >
       -- select a job -- 
    </option>
    <option>
      1
    </option>
    <option>
      4
    </option>
  </select>,
]
`;

exports[`Jobs Compare should update graph when graph updater updates value 1`] = `
<div>
  <div>
    <div
      class="row justify-content-md-center"
    >
      <div
        class="col-7"
      >
        <div
          class="comparer-dropdowns container"
        >
          <form
            class="row"
          >
            <div
              class="col-5"
            >
              <label
                for="job1"
              >
                Job 1
              </label>
              <select
                class="form-control"
                id="job1"
                name="job1"
                required=""
              >
                <option
                  disabled=""
                  value=""
                >
                   -- select a job -- 
                </option>
                <option>
                  1
                </option>
                <option>
                  2
                </option>
              </select>
            </div>
            <div
              class="col-5"
            >
              <label
                for="job2"
              >
                Job 2
              </label>
              <select
                class="form-control"
                id="job2"
                name="job2"
                required=""
              >
                <option
                  disabled=""
                  value=""
                >
                   -- select a job -- 
                </option>
                <option>
                  1
                </option>
                <option>
                  2
                </option>
              </select>
            </div>
            <div
              class="col-2 margin-top-auto"
            >
              <button
                class="btn btn-primary"
              >
                Compare
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-2"
      >
        <div
          class="comparer-config"
        >
          <div
            class="user-inputs"
          >
            <div
              class="form-row"
            >
              <h5>
                <b>
                  <label
                    for="series-filer"
                  >
                    Select line-graphs:
                  </label>
                </b>
              </h5>
              <div
                class=" css-2b097c-container"
              >
                <div
                  class=" css-yk16xz-control"
                >
                  <div
                    class=" css-g1d714-ValueContainer"
                  >
                    <div
                      class=" css-1wa3eu0-placeholder"
                    >
                      Select...
                    </div>
                    <div
                      class="css-b8ldur-Input"
                    >
                      <div
                        class=""
                        style="display: inline-block;"
                      >
                        <input
                          aria-autocomplete="list"
                          aria-label="tags"
                          autocapitalize="none"
                          autocomplete="off"
                          autocorrect="off"
                          id="series-filer"
                          spellcheck="false"
                          style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                          tabindex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class=" css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class=" css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class=" css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div>
                  <input
                    name="tags"
                    type="hidden"
                  />
                </div>
              </div>
            </div>
          </div>
          <form
            class="user-inputs"
          >
            <div
              class="form-row"
            >
              <h5>
                <b>
                  <label>
                    Select interventions:
                  </label>
                </b>
              </h5>
              <label
                for="job1-interventions-filter"
              >
                Job 1:
              </label>
              <div
                class=" css-2b097c-container"
              >
                <div
                  class=" css-yk16xz-control"
                >
                  <div
                    class=" css-g1d714-ValueContainer"
                  >
                    <div
                      class=" css-1wa3eu0-placeholder"
                    >
                      Select...
                    </div>
                    <div
                      class="css-b8ldur-Input"
                    >
                      <div
                        class=""
                        style="display: inline-block;"
                      >
                        <input
                          aria-autocomplete="list"
                          aria-label="tags"
                          autocapitalize="none"
                          autocomplete="off"
                          autocorrect="off"
                          id="job1-interventions-filter"
                          spellcheck="false"
                          style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                          tabindex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class=" css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class=" css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class=" css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div>
                  <input
                    name="job1-interventions"
                    type="hidden"
                  />
                </div>
              </div>
            </div>
            <div
              class="form-row"
            >
              <label
                for="job2-interventions-filter"
              >
                Job 2:
              </label>
              <div
                class=" css-2b097c-container"
              >
                <div
                  class=" css-yk16xz-control"
                >
                  <div
                    class=" css-g1d714-ValueContainer"
                  >
                    <div
                      class=" css-1wa3eu0-placeholder"
                    >
                      Select...
                    </div>
                    <div
                      class="css-b8ldur-Input"
                    >
                      <div
                        class=""
                        style="display: inline-block;"
                      >
                        <input
                          aria-autocomplete="list"
                          aria-label="tags"
                          autocapitalize="none"
                          autocomplete="off"
                          autocorrect="off"
                          id="job2-interventions-filter"
                          spellcheck="false"
                          style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                          tabindex="0"
                          type="text"
                          value=""
                        />
                        <div
                          style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class=" css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class=" css-1okebmr-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class=" css-tlfecz-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-6q0nyr-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <div>
                  <input
                    name="job2-interventions"
                    type="hidden"
                  />
                </div>
              </div>
            </div>
            <div
              class="form-row"
            >
              <button
                class="btn btn-primary"
              >
                Show Interventions
              </button>
            </div>
          </form>
        </div>
      </div>
      <div
        class="col-10"
      >
        <div
          class="jobs-compare-chart"
        >
          <div
            class="graph-wrap"
          >
            <div
              class="graph-actions"
            >
              <p
                class="action"
              >
                Zoom: 
                <code
                  class="control"
                >
                  click-drag
                </code>
              </p>
              <p
                class="action"
              >
                Pan: 
                <code
                  class="control"
                >
                  shift-click-drag
                </code>
              </p>
              <p
                class="action"
              >
                Restore: 
                <code
                  class="control"
                >
                  double-click
                </code>
              </p>
              <button
                class="btn btn-sm btn-secondary"
                disabled=""
              >
                Export graph data
              </button>
            </div>
            <div
              data-testid="visualization"
              id="vis"
              style="height: 320px; width: 480px;"
            >
              <div
                style="text-align: left; position: relative; width: 0px; height: 0px;"
              >
                <canvas
                  height="0"
                  style="position: absolute; width: 0px; height: 0px;"
                  width="0"
                />
                <canvas
                  height="0"
                  style="position: absolute; width: 0px; height: 0px;"
                  width="0"
                />
                <input
                  class="dygraph-roller"
                  size="2"
                  style="display: block; top: -63px; left: 57px;"
                  type="text"
                />
                <div
                  style="position: absolute; left: 0px; top: 0px; width: -5px; height: 28px; font-size: 20px;"
                >
                  <div
                    class="dygraph-label dygraph-title"
                  >
                    Time Series Graph
                  </div>
                </div>
                <div
                  style="position: absolute; left: 0px; top: -18px; width: -5px; height: 18px; font-size: 16px;"
                >
                  <div
                    class="dygraph-label dygraph-xlabel"
                  >
                    Hours
                  </div>
                </div>
                <div
                  style="position: absolute; left: 0px; top: 28px; width: 0px; height: -46px; font-size: 16px;"
                >
                  <div
                    class="dygraph-label-rotate-right"
                    style="position: absolute; width: -46px; height: 0px; top: -23px; left: 23px;"
                  >
                    <div
                      class="dygraph-label dygraph-ylabel"
                    >
                      Number of Agents
                    </div>
                  </div>
                </div>
                <div
                  class="dygraph-legend"
                  style="left: -6px; top: 28px;"
                >
                  <span
                    style="font-weight: bold; color: #0A1045;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #0A1045;"
                    />
                     susceptible_job1
                  </span>
                   
                  <span
                    style="font-weight: bold; color: #00C2D1;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #00C2D1;"
                    />
                     infected_job1
                  </span>
                   
                  <span
                    style="font-weight: bold; color: #F9E900;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #F9E900;"
                    />
                     hospitalized_job1
                  </span>
                   
                  <span
                    style="font-weight: bold; color: #F6AF65;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #F6AF65;"
                    />
                     recovered_job1
                  </span>
                   
                  <span
                    style="font-weight: bold; color: #ED33B9;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #ED33B9;"
                    />
                     deceased_job1
                  </span>
                   
                  <span
                    style="font-weight: bold; color: #8588a2;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #8588a2;"
                    />
                     susceptible_job2
                  </span>
                   
                  <span
                    style="font-weight: bold; color: #80e1e8;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #80e1e8;"
                    />
                     infected_job2
                  </span>
                   
                  <span
                    style="font-weight: bold; color: #fcf480;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #fcf480;"
                    />
                     hospitalized_job2
                  </span>
                   
                  <span
                    style="font-weight: bold; color: #fbd7b2;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #fbd7b2;"
                    />
                     recovered_job2
                  </span>
                   
                  <span
                    style="font-weight: bold; color: #f699dc;"
                  >
                    <div
                      class="dygraph-legend-line"
                      style="border-bottom-color: #f699dc;"
                    />
                     deceased_job2
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
